### ISSUE TITLE
Reusing Preconnections?

### ISSUE THREAD

**Issue by tfpauly** [ietf-tapswg/api-drafts#222]
Regarding this text:

```
The Initiate() Action consumes the Preconnection. Once Initiate() has
been called, no further properties may be added to the Preconnection, and
no subsequent establishment call may be made on the Preconnection.
```

This seems to imply that the Preconnection object, which may be full of many carefully tuned settings that could be useful for another Connection, is no longer usable. While I understand that the specific Preconnection cannot change for a Connection, this is inconvenient to the application.

I see two good options to solve this:

- Make a deep copy of the Preconnection upon Initiate(), and say that a Preconnection may be used or changed but changes can no longer influence the Connection (this is what Network.framework currently does)
- Allow a Preconnection.clone() method to let the applications do the deep copy themselves when and if they want to re-use.

**Comment by mirjak** [comment]
+1 to reusing the preconnection. Do you even need a deep copy. The connection is a new object that should be independent of the preconnection anyway, no?

Another thought: Are there actually ways to get different connections out of the preconenction. E.g. you do racing and end up with two open connections and (suddenly) want to use both...?

**Comment by theri** [comment]
I'd like to see this as well.
To me it makes most sense to allow calling Initiate() on the same Preconnection multiple times to get multiple Connections, but not allow to change this Preconnection anymore after the first Initiate. This way it would be similar to Listen() and Rendezvous(), where we have a Preconnection that yields multiple Connections, and that cannot be changed anymore after Listen() or Rendezvous(). So here we already have a binding between the Preconnection and the Connections that come out of it.

If the application wants to change the Preconnection before re-using it, I'd be in favor of letting it explicitly clone the Preconnection - not sure if we actually want to call it clone() though, because we already have Connection.Clone(), which creates a Connection Group, a completely different thing.

**Comment by mirjak** [comment]
Why do you think it should not be allowed that a preconnection can be changed after the first initiate? If you don't change it you should get the same kind of connection, if you change it you might get something different. Your choice!

**Comment by theri** [comment]
I wanted the API contract to be similar for different forms of Connection establishment. But I may be overthinking this.

**Comment by philsbln** [comment]
First, I like the idea of reusing Preconnections.

I am not sure whether we should define now whether they should be changeable after the first initiation. There might be kinds of Preconnections where this is a good idea, there might be others where it is not. 
I anticipate one can realize connection pools as a sub-class of Preconnection. In this case, in such cases, it should not be possible to change a Preconnection.

**Comment by mirjak** [comment]
So the difference is if you think you morph a preconnection into a connection, or if the preconnection create the connection. i think the first think does not make sense because their are two completely different things; they have not actions or events in common.

**Comment by philsbln** [comment]
@mirjak The morphing one has its beauty for modeling TCP fastopen / 0-RTT data.

**Comment by csperkins** [comment]
Whether it makes sense to re-use a `Preconnection` perhaps depends on how fully specified it is? Reusing a completely specified `Preconnection`, that will just re-open an identical 5-tuple as an existing `Connection` might not make sense. Reusing a partially specified `Preconnection`, e.g., with the source port left as a wildcard to new connections can open without conflicts, seems reasonable.

**Comment by tfpauly** [comment]
@csperkins right, using the exact same set of fully-specified properties isn't going to work. But having a partially specified one and copying it, or modifying one that was used (by switching out the Remote, etc) would be useful.

**Comment by mirjak** [comment]
So i think there is actually not that much too deep copy because a preconception is not a super class of a connection. At the moment when you call initiate or rendezvous the preconception should create a connection and store all information regarding this connection in the connection object which is later returned. The only thing that needs "copying" is the properties but you need to create a new object anyway because preconnection properties have preferences and connection properties not but have to be marked as mutable or unmutable. These two things are also not the same. So the only that that's left to copy maybe are the endpoints. Not sure if the preconnection and connection both have a copy of the endpoints or if an endpoint only exists once and both have a reference (which means if the endpoint gets somehow updated that's relevant for both) or maybe even the connection does not need a reference to the endpoint anymore...?

**Comment by philsbln** [comment]
I have no problem with an Interface where some Preconnections can be reused and others can not.

That makes a pretty nice case for ``Initialte()``-- preconnection can be reused --  and ``listen()`` -- preconnection reuses itself for multiple incoming connections.
For ``rendezvous()`` I see this API-wise as some kind of listen called at both sides.

**Comment by tfpauly** [comment]
Side meeting says we can reuse the initiate call. Implementation should specify that a deep copy is made on Initiate() or Listen()

**Comment by philsbln** [comment]
Will be trivial to fix after #301 is done

**Comment by theri** [comment]
Sorry to reopen an old issue, but when reading the current API draft, I noticed that it still says that reusing the Preconnection is possible for Listeners, but not possible for Initiate() and Rendezvous(). For example, see Section 6.1: "The Initiate() Action consumes the Preconnection […]"

From what I read in this issue, I think Preconnections should be allowed to be reused for Initiate() and Rendezvous() as well (possibly with a note on deep-copy).
